activerecord/lib/active_record/validations/associated.rb

```diff
@@ -11,6 +16,12 @@ def validate_each(record, attribute, value)
 
       private
         def valid_object?(record)
+          if @associated_validations
+            @associated_validations.arity.zero? ?
+              record.singleton_class.instance_exec(&@associated_validations) :
+              @associated_validations.call(record.singleton_class)
+          end
```

Comment:
@jonathanhefner this feels related to https://github.com/rails/rails/pull/42907 in its goal. With the presence of something like `CompositeValidator`, this implementation might require less magic and `singleton_class` manipulation.